mscore/plugin/api/playevent.h

```diff
@@ -0,0 +1,151 @@
+//=============================================================================
+//  MuseScore
+//  Music Composition & Notation
+//
+//  Copyright (C) 2019 Werner Schweer and others
+//
+//  This program is free software; you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License version 2
+//  as published by the Free Software Foundation and appearing in
+//  the file LICENCE.GPL
+//=============================================================================
+
+#ifndef __PLUGIN_API_PLAYEVENT_H__
+#define __PLUGIN_API_PLAYEVENT_H__
+
+#include "libmscore/noteevent.h"
+#include "elements.h"
+
+namespace Ms {
+namespace PluginAPI {
+
+class Element;
+class Note;
+
+extern Ms::Element* unwrap(Element* e);
+
+//---------------------------------------------------------
+//   PlayEvent
+//    Wrapper class for internal Ms::NoteEvent
+//
+//   This is based on the wrapper in excerpt.h.
+//---------------------------------------------------------
+
+class PlayEvent : public QObject {
+      Q_OBJECT
+      /// The relative pitch of the event.
+      /// This is added to the parent note's actual pitch.
+      /// \since MuseScore 3.3
+      Q_PROPERTY(int pitch READ pitch WRITE setPitch)
+      /// Time to turn on the note event.
+      /// \since MuseScore 3.3
+      Q_PROPERTY(int ontime READ ontime WRITE setOntime)
+      /// The length of time for the event.
+      /// \since MuseScore 3.3
+      Q_PROPERTY(int len READ len WRITE setLen)
+      /// Time note will turn off.
+      /// This value derived from onTime and len.
+      /// \since MuseScore 3.3
+      Q_PROPERTY(int offtime READ offtime)
+      /// \cond MS_INTERNAL
+
+   protected:
+      Ms::NoteEvent* ne;
+      QObject* parentNote;
+
+   public:
+
+      PlayEvent(Ms::NoteEvent* _ne = new Ms::NoteEvent(), QObject* _parent = nullptr)
+         : QObject(), ne(_ne), parentNote(_parent) {}
+      // Delete the NoteEvent if parentless.
+      virtual ~PlayEvent() { if (parentNote == nullptr) delete ne; }
+
+      const Ms::NoteEvent& getNoteEvent() { return *ne; }
+      void setParentNote(QObject* parent) { this->parentNote = parent; }
+      Note* note() { return reinterpret_cast<Note*>(parentNote); }
+
+      int pitch() const { return ne->pitch(); }
+      int ontime() const { return ne->ontime(); }
+      int offtime() const { return ne->offtime(); }
+      int len() const { return ne->len(); }
+      void setPitch(int v);
+      void setOntime(int v);
+      void setLen(int v);
+      /// \endcond
+};
+
+//---------------------------------------------------------
+//   wrap
+///   \cond PLUGIN_API \private \endcond
+///   \relates Ms::NoteEvent
+//---------------------------------------------------------
+
+template <class Wrapper, class T>
+Wrapper* playEventWrap(T* t, QObject* parent)
```

Comment:
Isn't this function intended to be used only for `PlayEvent`s? If so then it should be not necessary to make it a template.

Reply:
I tried dispensing with the template as I totally agree with you. It just seems wrong.

I could not get Qt to expose a non-templated version from QML. 

As some point I just went with the pattern that worked for excerpt.h as I had spend too much time trying the other approach. I was at the point of: "Am I doing this to add a feature versus am I being pure enough." After a couple days I was ready to walk from the work in frustration so I un-stashed this code and went with it.

What would really be great if there was a MuseScore place I could go when run into these issues against experienced folks to get more eyes on it. The PR is not the place and Google-ing only gets a person so far. Many times it's situational to the target program and those people might say: "Oh yeah, that happened to me. Here's how I handled it!"

BTW, Your observation is also true of the implementation I borrowed from: excerpt.h's.
It specialized to a single use specialization... but it worked.

-Dale

Reply:
I think once the little bits are settled I'll take one more whack at the non-template version under the assumption that other things may have gotten in my way.

It just bugs the hell out of me to use the template in this way.

-Dale

Reply:
I've taken care of the other inspection comments and this one is WIP.

Reply:
I've just pushed up a non-templated version of the wrapper code stuff for PlayEvents.

I turns out it's easier to do things when you have working code to start with.

-Dale